### DISCUSSION TITLE
MariaDB deprecation

### DISCUSSION THREAD

**Discussion by westhouseit** [craftcms/cms#15851] [General]
I'm a developer supporting clients who build sites in CraftCMS. With the amount of money spent on the CMS and Commerce I'm surprised to find no detailed explanation of the MariaDB deprecation. MariaDB is the default database system and sometimes the only option across many hosts.

The only details I've found so far are in these links, so perhaps there's a technical discussion somewhere else?
https://www.craftcms.com/knowledge-base/craft-database-options-and-best-practices
https://craftcms.com/docs/5.x/requirements.html

It's rather puzzling to read comments from yourselves like
> Due to its diverging parity with MySQL, we no longer recommend MariaDB.

Postgres is much further from MySQL, but you support it. Does Pixel & Tonic have an issue with the MariaDB project? I can't find any posts on the MariaDB project regarding Craft. Perhaps you could discuss your technical issues with the MariaDB developers, I'm sure they'd be happy to help you.

And this comment reads like it was written by marketing: 
>Due to recent departures from the MySQL spec, we no longer recommend MariaDB for use with Craft.

Postgres never matched MySQL, nor does it or any other product need to do so. There's certainly a SQL spec/Standard, but what's a MySQL spec? If you're referring to the mysql driver, are you having issues using it with MariaDB?

Another comment of yours
>we no longer recommend MariaDB for sites with many users or a large volume of content.

directly contradicts Amazon
>MariaDB is more scalable and offers a higher query speed when compared to MySQL. This makes it good for managing large-sized data. You will also find more features in MariaDB that MySQL doesn’t have, like sequence storage engines and virtual columns. You can also use multiple engines in one table.
https://aws.amazon.com/compare/the-difference-between-mariadb-vs-mysql

P.S. WordPress offers full support for MariaDB.

**Comment by thupsi** [comment]
Take a look at https://github.com/craftcms/cms/discussions/13997

**Reply by westhouseit** [reply]
Yes, I asked for more information there, but none is forthcoming.

Brandon mentions "numerous performance and stability issues with [MariaDB]", but I can't find a single issue from P&T here: https://jira.mariadb.org/issues/?jql=text%20~%20%22craftcms%22
In any case MySQL does NOT "fully support JSON"
> MySQL and MariaDB also do not both support all JSON functions. MariaDB supports JSON_QUERY and JSON_EXISTS, while MySQL doesn’t. Similarly, MySQL supports the native JSON data type JSON_TABLE that MariaDB does not.

Even that comment from Amazon is out of date since MariaDB added JSON_TABLE support in 10.6.0.

If the real issue is that the P&T elected to use a function that isn't supported by MariaDB 11.4 then why not just come out and say it?

**Comment by brandonkelly** [comment]
We never set out to support MariaDB to begin with. MariaDB gained popularity for being a “drop-in replacement” for MySQL at a time when MySQL’s future was in doubt. Instead, MySQL development is thriving, and MariaDB has given up on their stated goal (likely in part because the company behind it is [in shambles](https://medium.com/@imashadowphantom/mariadb-com-is-dead-long-live-mariadb-org-b8a0ca50a637)). We have been seeing increasing performance issues with MariaDB for a couple years now, and decided to stop recommending it as a result of that. Though support for MariaDB is not _deprecated_ – we will continue supporting it as long as we can! If you have existing customers that aren’t having problems with it, feel free to continue using it.

**Reply by westhouseit** [reply]
Perhaps, but then you didn't support Postgres at the time either. Though P&T have officially supported MariaDB since Craft v3: https://craftcms.com/docs/3.x/requirements.html
DB support via the built-in drivers for the underlying framework that you use would be recommended since they support MariaDB.

That post on Medium is very suspect - an unreliable platform, with a post by an anonymous person. I'm not saying they're wrong, but it's a bad source of information. Plus, should we write off an open-source product because of a bad company?

Also, you might not have used the word "deprecated", but the meaning was there. 
A quick search returns this: "Deprecation is the discouragement of use of something human-made, such as a term, feature, design, or practice. Typically something is deprecated because it is claimed to be inferior compared to other options available."
"when something is available or allowed, but is not recommended or that its failings are recognized"


In any case my original points were mostly ignored.

**Reply by brandonkelly** [reply]
In software, “deprecated” usually means that something is not only not recommended, but expected to be removed entirely down the road. Which again, is not the case here.

> In any case my original points were mostly ignored.

We don’t have the bandwidth or expertise to help maintain MariaDB, including reporting obscure performance issues. What we know is that we’ve seen a pattern of performance issues that tend to be solved when the customer switches from MariaDB to MySQL, so we’ve chosen to not recommend Maria. Simple as that.

**Reply by westhouseit** [reply]
> What we know is that we’ve seen a pattern of performance issues

Then where is the help article or blog post detailing this? Issues with one of the most popular databases used by your customers should be clearly outlined, not hidden in closed support threads.

Following your reasoning I should advise my clients that, although CraftCMS isn't deprecated, we recommend against using it due to performance issues and divergence from industry standards. We also won't help Pixel & Tonic fix issues by reporting anything helpful since no one pays us to do that. Does that seem reasonable to you?

**Reply by brandonkelly** [reply]
We’re just talking in circles so I’m locking the discussion. Feel free to continue using MariaDB if you’re happy with it; we will continue to support you in doing so. There will just be limits to what we can help you with, if you’re facing a performance issue that turns out to be due to MariaDB.
